### ISSUE TITLE
test: better error message for panicked test case

### ISSUE THREAD

**Issue by xxchan** [risingwavelabs/risingwave#4827]
We can’t use panic hook to do things like checking which test case panicked, since panic hook is global and tests may run in parallel. 

For test binaries build with `libtest-mimic` (`sqlsmith` & `run-planner-test`), one solution is to print necessary information beforehand (as we refactored in #4800, #4228). Then output will only be displayed when a test fails/panics.

A limitation is that if a test case is too large, there may be many irrelevant messages. e.g., a planner test case (a yaml file) has tens of SQLs (BTW, for this problem, maybe we can use `TestCase::run` as test unit, instead of `run_test_file`?). 

Another limitation is that we can't do similar things for `apply-planner-test` yet, since it's not a test binary. 

Maybe we can:
1. Use one thread for each test, and then use `thread_local` to store test progress and use panic hook to display it.
2. `catch_unwind`. Actually `libtest-mimic` uses `catch_unwind`.

cc @skyzh

**Comment by xxchan** [comment]
Won't fix now since it's not very urgent.
